### DISCUSSION TITLE
How to get all changed components form all entities?

### DISCUSSION THREAD

**Discussion by codpe** [bevyengine/bevy#10607] [Q&A]
I am building game server using headless bevy ecs.
I want to automatically synchronize all changes from large amount of  components and entities to all clients.
Components that should be synchronized will be also serializable with protobuf.   
The best solution would be one that can effectively iterate only trough changed components. Ideally if it is universal enough, that when adding new types of components, it requires no or little code.
Also I send sync data only to players that are close enough, so there is a rdtree for that.

Of course I could write a system for each type of component with
```
Query<(Entity, Position, MyComponent1), Changed<MyComponent1>
```
but it intuitively it seems no the right solution and then each such system should search rdtree for players that are close enough for components that are from entity with same Position.  

What would be the best approach to do this?

**Comment by rlidwka** [comment]
There is a way to get `&mut World`, and iterate through all entities and components like Bevy engine would. But I suspect this is probably overkill here.

> then each such system should search rdtree for players that are close enough

Can you have one system that finds all "close enough" players (and sets marker component on each), then have a system per each type of component that syncs that particular component for only close enough players?

**Reply by codpe** [reply]
> There is a way to get &mut World, and iterate through all entities and components like Bevy engine would. But I suspect this is probably overkill here.

In the end I will probably will have >100`000 entities and 100+ component types, but only some 200 - 1000 components changed in each tick.

> Can you have one system that finds all "close enough" players (and sets marker component on each), then have a system per each type of component that syncs that particular component for only close enough players?

Problem is that I need to find close players list for each entity that has changed. With this approach I could add ClosePlayers component for each entity when iterating over Changed components, but then again I will have to check, if that entity already have ClosePlayers calculated. With this approach I could even optimize recalculation of ClosePlyars if non of players and entity itself has not changed its position a lot.

Using your suggestion, i came to something like this
```rs
fn check_changes<T : Component>(mut query: Query<(&ID, &Position, &T, &ClosePlayers), Changed<T>>, kdtree: Res<PlayerKDTree>) {
  for (id, position, comp, close_players) in query.iter_mut() {
    if !close_players.has {
      close_players.players = kdtree.getClosePlayers(position);
      close_players.has = true;
    }
    for player in close_players.players.iter_mut() {
      player.addChangedComponent(id, comp);
    }    
  }
}
```
and i will have to add system for each component type I want to sync with clients
```rs
.add_systems(Update, check_changes::<Position>)
.add_systems(Update, check_changes::<Health>)
```
but I am not sure isn't there a fundamentally better solution and is this a good practice to have 100+ such systems that search for changed components, when most of those queries will be empty?
